ext/standard/random.c

```diff
@@ -85,26 +89,75 @@ static int php_random_bytes(void *bytes, size_t size)
 	}
 #elif HAVE_DECL_ARC4RANDOM_BUF
 	arc4random_buf(bytes, size);
+#elif HAVE_DECL_GETRANDOM
+	/* Linux getrandom(2) syscall */
+	size_t read_bytes = 0;
+	size_t amount_to_read = 0;
+	ssize_t n;
+
+	/* Keep reading until we get enough entropy */
+	do {
+		amount_to_read = size - read_bytes;
+		/* Below, (bytes + read_bytes)  is pointer arithmetic.
+
+		   bytes   read_bytes  size
+		     |      |           |
+		    [#######=============] (we're going to write over the = region)
+		             \\\\\\\\\\\\\
+		              amount_to_read
+
+		*/
+		n = getrandom(bytes + read_bytes, amount_to_read, 0);
+
+		if (n == -1) {
+			if (errno == EINTR || errno == EAGAIN) {
+				/* Try again */
+				continue;
+			}
+			/*
+				If the syscall fails, we are doomed. The loop that calls
+				php_random_bytes should be terminated by the exception instead
+				of proceeding to demand more entropy.
+			*/
+			zend_throw_exception(zend_ce_exception, "Could not gather sufficient random data", errno);
+			return FAILURE;
+		}
+
+		read_bytes += (size_t) n;
+	} while (read_bytes < size);
 #else
 	int    fd = RANDOM_G(fd);
+	struct stat st;
 	size_t read_bytes = 0;
+	ssize_t n;
 
 	if (fd < 0) {
-#if HAVE_DEV_ARANDOM
-		fd = open("/dev/arandom", O_RDONLY);
-#elif HAVE_DEV_URANDOM
+#if HAVE_DEV_URANDOM
 		fd = open("/dev/urandom", O_RDONLY);
 #endif
 		if (fd < 0) {
 			zend_throw_exception(zend_ce_exception, "Cannot open source device", 0);
 			return FAILURE;
 		}
-
+		/* Does the file exist and is it a character device? */
+		if (fstat(fd, &st) != 0 || !S_ISCHR(st.st_mode)) {
+			close(fd);
+			zend_throw_exception(zend_ce_exception, "Error reading from source device", 0);
+			return FAILURE;
+		}
+#ifdef __linux__
+		// Make sure that /dev/urandom is the proper urandom device on Linux
+		if (st.st_rdev != makedev(1, 9)) {
```

Comment:
I don't like hardcoding this. Does it come from some standard that /dev/urandom on Linux always must be the same? Why we need to check this anyway? If someone has permissions to mess with /dev to fool us, the system has way more problems than PHP could fix. Checking if the file has been opened successfully, e.g. to see if we aren't in broken chroot() env or broken permissions - that's fine, but beyond that I don't think we should bother, and especially hardcode device numbers inside the code.